engine/src/main/java/org/terasology/rendering/animation/Animation.java

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright 2016 MovingBlocks
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.terasology.rendering.animation;
+
+import org.terasology.assets.AssetData;
+import org.terasology.entitySystem.Component;
+import org.terasology.entitySystem.entity.EntityRef;
+
+import java.util.Collections;
+import java.util.Set;
+import java.util.List;
+import java.util.ArrayList;
+
+/*
+ */
+public class Animation implements Component, AssetData {
+    private List<AnimationListener> listeners;
+    private List<Frame> frames;
+    private int repeat;
+
+    private float elapsedTime;
+    private int currentFrame;
+    private int currentRepeatCount;
+
+    private enum AnimState {
+        PRESTART, PAUSED, RUNNING, FINISHED
+    }
+    private AnimState currentState;
+
+    public Animation() {
+        listeners = new ArrayList<AnimationListener>();
+        frames = new ArrayList<Frame>();
+        currentState = AnimState.PRESTART;
+        repeat = 0;
+        elapsedTime = 0;
+        currentFrame = 0;
+    }
+
+    public void addFrame(Frame frame) {
+        this.frames.add(frame);
+    }
+
+    public void setRepeat(int repeat) {
+        this.repeat = repeat;
+    }
+
+    public void update(float delta) {
+        switch (this.currentState) {
+        case RUNNING: {
+            frames.get(currentFrame).update(delta);
+            if (frames.get(currentFrame).isFinished()) {
+                currentFrame++;
+                if (currentFrame >= frames.size()) {
+                    currentRepeatCount++;
+                    if (currentRepeatCount >= repeat ||
+                        currentRepeatCount >= -(repeat + 1)) {
```

Comment:
I should write the documentation for it. 0 means repeat forever, -1 means repeat and inverse forever, and anything >= 1 means repeat that many times, and anything <= -2 means repeat and inverse that many times.